File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/repository/memory/MemoryRepository.java
package com.androidcollider.easyfin.common.repository.memory;

import android.util.Log;

import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager;
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager;
import com.androidcollider.easyfin.common.models.Account;
import com.androidcollider.easyfin.common.models.DateConstants;
import com.androidcollider.easyfin.common.models.Debt;
import com.androidcollider.easyfin.common.models.Rates;
import com.androidcollider.easyfin.common.models.Transaction;
import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.repository.Repository;
import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Flowable;

/**
 * @author Ihor Bilous
 */

public class MemoryRepository implements Repository {

    private static final String TAG = MemoryRepository.class.getSimpleName();

    private NumberFormatManager numberFormatManager;
    private List<Account> accountList;
    private List<Transaction> transactionList;
    private List<Debt> debtList;
    private double[] ratesArray;
    private String[] currencyArray;
    private List<TransactionCategory> transactionCategoryIncomeList;
    private List<TransactionCategory> transactionCategoryExpenseList;


    public MemoryRepository(NumberFormatManager numberFormatManager,
                            ResourcesManager resourcesManager) {
        this.numberFormatManager = numberFormatManager;
        currencyArray = resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY);
    }

    @Override
    public Flowable<Account> addNewAccount(Account account) {
        return Flowable.fromCallable(() -> {
            accountList.add(account);
            return account;
        });
    }

    @Override
    public Flowable<List<Account>> getAllAccounts() {
        Log.d(TAG, "getAllAccounts");
        return accountList == null ? null : Flowable.just(accountList);
    }

    @Override
    public Flowable<Account> updateAccount(Account account) {
        return Flowable.fromCallable(() -> {
            int pos = accountList.indexOf(account);
            return pos >= 0 ? accountList.set(pos, account) : null;
        });
    }

    @Override
    public Flowable<Boolean> deleteAccount(int id) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(id);
            boolean b = pos != -1;
            if (b) accountList.remove(pos);
            return b;
        });
    }

    @Override
    public Flowable<Boolean> transferBTWAccounts(int idAccount1, double accountAmount1, int idAccount2, double accountAmount2) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(idAccount1);
            boolean b1 = pos != -1;
            if (b1) {
                accountList.get(pos).setAmount(accountAmount1);
            }
            pos = getAccountPosById(idAccount2);
            boolean b2 = pos != -1;
            if (b2) {
                accountList.get(pos).setAmount(accountAmount2);
            }
            return b1 && b2;
        });
    }

    @Override
    public Flowable<Transaction> addNewTransaction(Transaction transaction) {
        return Flowable.fromCallable(() -> {
            transactionList.add(0, transaction);
            int pos = getAccountPosById(transaction.getIdAccount());
            if (pos != -1) {
                accountList.get(pos).setAmount(transaction.getAccountAmount());
            }
            return transaction;
        });
    }

    @Override
    public Flowable<List<Transaction>> getAllTransactions() {
        Log.d(TAG, "getAllTransactions");
        return transactionList == null ? null : Flowable.just(transactionList);
    }

    @Override
    public Flowable<Transaction> updateTransaction(Transaction transaction) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(transaction.getIdAccount());
            if (pos != -1) {
                accountList.get(pos).setAmount(transaction.getAccountAmount());
            }
            pos = transactionList.indexOf(transaction);
            return pos >= 0 ? transactionList.set(pos, transaction) : null;
        });
    }

    @Override
    public Flowable<Boolean> updateTransactionDifferentAccounts(Transaction transaction, double oldAccountAmount, int oldAccountId) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(transaction.getIdAccount());
            boolean b1 = pos != -1;
            if (b1) {
                accountList.get(pos).setAmount(transaction.getAccountAmount());
            }
            pos = getAccountPosById(oldAccountId);
            boolean b2 = pos != -1;
            if (b2) {
                accountList.get(pos).setAmount(oldAccountAmount);
            }
            pos = transactionList.indexOf(transaction);
            boolean b3 = pos != -1 && transactionList.set(pos, transaction) != null;
            return b1 && b2 && b3;
        });
    }

    @Override
    public Flowable<Boolean> deleteTransaction(int idAccount, int idTransaction, double amount) {
        return Flowable.fromCallable(() -> {
            int pos = getTransactionPosById(idTransaction);
            boolean b = pos != -1;
            if (b) {
                transactionList.remove(pos);
                pos = getAccountPosById(idAccount);
                b = pos != -1;
                if (b) {
                    accountList.get(pos).setAmount(accountList.get(pos).getAmount() - amount);
                }
            }
            return b;
        });
    }

    @Override
    public Flowable<Debt> addNewDebt(Debt debt) {
        return Flowable.fromCallable(() -> {
            debtList.add(debt);
            int pos = getAccountPosById(debt.getIdAccount());
            if (pos != -1) {
                accountList.get(pos).setAmount(debt.getAccountAmount());
            }
            return debt;
        });
    }

    @Override
    public Flowable<List<Debt>> getAllDebts() {
        Log.d(TAG, "getAllDebts");
        return debtList == null ? null : Flowable.just(debtList);
    }

    @Override
    public Flowable<Debt> updateDebt(Debt debt) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(debt.getIdAccount());
            if (pos != -1) {
                accountList.get(pos).setAmount(debt.getAccountAmount());
            }
            pos = debtList.indexOf(debt);
            return pos >= 0 ? debtList.set(pos, debt) : null;
        });
    }

    @Override
    public Flowable<Boolean> updateDebtDifferentAccounts(Debt debt, double oldAccountAmount, int oldAccountId) {
        return Flowable.fromCallable(() -> {
            int pos = getAccountPosById(debt.getIdAccount());
            boolean b1 = pos != -1;
            if (b1) {
                accountList.get(pos).setAmount(debt.getAccountAmount());
            }
            pos = getAccountPosById(oldAccountId);
            boolean b2 = pos != -1;
            if (b2) {
                accountList.get(pos).setAmount(oldAccountAmount);
            }
            pos = debtList.indexOf(debt);
            boolean b3 = pos != -1 && debtList.set(pos, debt) != null;
            return b1 && b2 && b3;
        });
    }

    @Override
    public Flowable<Boolean> deleteDebt(int idAccount, int idDebt, double amount, int type) {
        return Flowable.fromCallable(() -> {
            int pos = getDebtPosById(idDebt);
            boolean b = pos != -1;
            if (b) {
                debtList.remove(pos);
                pos = getAccountPosById(idAccount);
                b = pos != -1;
                if (b) {
                    accountList.get(pos).setAmount(type == 1 ?
                            accountList.get(pos).getAmount() - amount :
                            accountList.get(pos).getAmount() + amount
                    );
                }
            }
            return b;
        });
    }

    @Override
    public Flowable<Boolean> payFullDebt(int idAccount, double accountAmount, int idDebt) {
        return Flowable.fromCallable(() -> {
            int pos = getDebtPosById(idDebt);
            boolean b = pos != -1;
            if (b) {
                debtList.remove(pos);
                pos = getAccountPosById(idAccount);
                b = pos != -1;
                if (b) {
                    accountList.get(pos).setAmount(accountAmount);
                }
            }
            return b;
        });
    }

    @Override
    public Flowable<Boolean> payPartOfDebt(int idAccount, double accountAmount, int idDebt, double debtAmount) {
        return Flowable.fromCallable(() -> {
            int pos = getDebtPosById(idDebt);
            boolean b = pos != -1;
            if (b) {
                debtList.get(pos).setAmountCurrent(debtAmount);
                pos = getAccountPosById(idAccount);
                b = pos != -1;
                if (b) {
                    accountList.get(pos).setAmount(accountAmount);
                }
            }
            return b;
        });
    }

    @Override
    public Flowable<Boolean> takeMoreDebt(int idAccount, double accountAmount, int idDebt, double debtAmount, double debtAllAmount) {
        return Flowable.fromCallable(() -> {
            int pos = getDebtPosById(idDebt);
            boolean b = pos != -1;
            if (b) {
                debtList.get(pos).setAmountCurrent(debtAmount);
                debtList.get(pos).setAmountAll(debtAllAmount);
                pos = getAccountPosById(idAccount);
                b = pos != -1;
                if (b) {
                    accountList.get(pos).setAmount(accountAmount);
                }
            }
            return b;
        });
    }

    @Override
    public Flowable<Map<String, double[]>> getTransactionsStatistic(int position) {
        return transactionList == null ?
                null :
                Flowable.fromCallable(() -> getTransactionsStatisticByPosition(position));
    }

    @Override
    public Flowable<Map<String, double[]>> getAccountsAmountSumGroupByTypeAndCurrency() {
        return accountList == null || debtList == null ?
                null :
                Flowable.fromCallable(this::getAccountsSumGroupByTypeAndCurrency);
    }

    @Override
    public Flowable<Boolean> updateRates(List<Rates> ratesList) {
        return Flowable.fromCallable(() -> {
            for (int i = 0; i < ratesArray.length; i++) {
                ratesArray[i] = ratesList.get(i).getAsk();
            }
            return true;
        });
    }

    @Override
    public Flowable<double[]> getRates() {
        Log.d(TAG, "getRates");
        return ratesArray == null ? null : Flowable.just(ratesArray);
    }

    @Override
    public Flowable<Boolean> setAllAccounts(List<Account> accountList) {
        return Flowable.fromCallable(() -> {
            this.accountList = new ArrayList<>();
            this.accountList.addAll(accountList);
            return true;
        });
    }

    @Override
    public Flowable<Boolean> setAllTransactions(List<Transaction> transactionList) {
        return Flowable.fromCallable(() -> {
            this.transactionList = new ArrayList<>();
            this.transactionList.addAll(transactionList);
            return true;
        });
    }

    @Override
    public Flowable<Boolean> setAllDebts(List<Debt> debtList) {
        return Flowable.fromCallable(() -> {
            this.debtList = new ArrayList<>();
            this.debtList.addAll(debtList);
            return true;
        });
    }

    @Override
    public Flowable<Boolean> setRates(double[] rates) {
        return Flowable.fromCallable(() -> {
            this.ratesArray = new double[4];
            System.arraycopy(rates, 0, this.ratesArray, 0, rates.length);
            return true;
        });
    }

    @Override
    public Flowable<TransactionCategory> addNewTransactionIncomeCategory(TransactionCategory transactionCategory) {
        return Flowable.fromCallable(() -> {
            transactionCategoryIncomeList.add(transactionCategory);
            return transactionCategory;
        });
    }

    @Override
    public Flowable<List<TransactionCategory>> getAllTransactionIncomeCategories() {
        return transactionCategoryIncomeList == null ? null : Flowable.just(transactionCategoryIncomeList);
    }

    @Override
    public Flowable<TransactionCategory> updateTransactionIncomeCategory(TransactionCategory transactionCategory) {
        return Flowable.fromCallable(() -> {
            int pos = transactionCategoryIncomeList.indexOf(transactionCategory);
            return pos >= 0 ? transactionCategoryIncomeList.set(pos, transactionCategory) : null;
        });
    }

    @Override
    public Flowable<Boolean> deleteTransactionIncomeCategory(int id) {
        return Flowable.fromCallable(() -> {
            int pos = getTransactionCategoryIncomePosById(id);
            boolean b = pos != -1;
            if (b) transactionCategoryIncomeList.remove(pos);
            return b;
        });
    }

    @Override
    public Flowable<Boolean> setAllTransactionIncomeCategories(List<TransactionCategory> transactionCategoryList) {
        return Flowable.fromCallable(() -> {
            this.transactionCategoryIncomeList = new ArrayList<>();
            this.transactionCategoryIncomeList.addAll(transactionCategoryList);
            return true;
        });
    }

    @Override
    public Flowable<TransactionCategory> addNewTransactionExpenseCategory(TransactionCategory transactionCategory) {
        return Flowable.fromCallable(() -> {
            transactionCategoryExpenseList.add(transactionCategory);
            return transactionCategory;
        });
    }

    @Override
    public Flowable<List<TransactionCategory>> getAllTransactionExpenseCategories() {
        return transactionCategoryExpenseList == null ? null : Flowable.just(transactionCategoryExpenseList);
    }

    @Override
    public Flowable<TransactionCategory> updateTransactionExpenseCategory(TransactionCategory transactionCategory) {
        return Flowable.fromCallable(() -> {
            int pos = transactionCategoryExpenseList.indexOf(transactionCategory);
            return pos >= 0 ? transactionCategoryExpenseList.set(pos, transactionCategory) : null;
        });
    }

    @Override
    public Flowable<Boolean> deleteTransactionExpenseCategory(int id) {
        return Flowable.fromCallable(() -> {
            int pos = getTransactionCategoryExpensePosById(id);
            boolean b = pos != -1;
            if (b) transactionCategoryExpenseList.remove(pos);
            return b;
        });
    }

    @Override
    public Flowable<Boolean> setAllTransactionExpenseCategories(List<TransactionCategory> transactionCategoryList) {
        return Flowable.fromCallable(() -> {
            this.transactionCategoryExpenseList = new ArrayList<>();
            this.transactionCategoryExpenseList.addAll(transactionCategoryList);
            return true;
        });
    }

    private Map<String, double[]> getAccountsSumGroupByTypeAndCurrency() {
        Log.d(TAG, "getAccountsSumGroupByTypeAndCurrency");
        Map<String, double[]> results = new HashMap<>();

        for (String currency : currencyArray) {
            List<Account> accountListFilteredByCurrency =
                    Stream.of(accountList)
                            .filter(a -> a.getCurrency().equals(currency))
                            .collect(Collectors.toList());
            double[] result = new double[4];
            double accountSum;
            for (int i = 0; i < 3; i++) {
                final int type = i;
                accountSum = 0;
                List<Account> accountListFilteredByType =
                        Stream.of(accountListFilteredByCurrency)
                                .filter(a -> a.getType() == type)
                                .collect(Collectors.toList());

                for (Account account : accountListFilteredByType) {
                    accountSum += account.getAmount();
                }
                result[i] = accountSum;
            }

            List<Debt> debtListFilteredByCurrency =
                    Stream.of(debtList)
                            .filter(d -> d.getCurrency().equals(currency))
                            .collect(Collectors.toList());

            double debtSum = 0;
            double debtVal;

            for (Debt debt : debtListFilteredByCurrency) {
                debtVal = debt.getAmountCurrent();
                if (debt.getType() == 1) {
                    debtVal *= -1;
                }
                debtSum += debtVal;
            }

            result[3] = debtSum;
            results.put(currency, result);
        }
        return results;
    }

    private Map<String, double[]> getTransactionsStatisticByPosition(int position) {
        Log.d(TAG, "getTransactionsStatisticByPosition");
        long period = 0;
        switch (position) {
            case 1:
                period = DateConstants.DAY;
                break;
            case 2:
                period = DateConstants.WEEK;
                break;
            case 3:
                period = DateConstants.MONTH;
                break;
            case 4:
                period = DateConstants.YEAR;
                break;
            case 5:
                period = Long.MAX_VALUE;
                break;
        }

        Map<String, double[]> result = new HashMap<>();
        long currentTime = System.currentTimeMillis();
        double cost, income;

        for (String currency : currencyArray) {
            List<Transaction> transactionListFilteredByCurrency =
                    Stream.of(transactionList)
                            .filter(t -> t.getCurrency().equals(currency))
                            .collect(Collectors.toList());

            cost = 0;
            income = 0;

            for (Transaction transaction : transactionListFilteredByCurrency) {
                long date = transaction.getDate();
                double amount = transaction.getAmount();
                if (currentTime > date && period >= (currentTime - date)) {
                    if (numberFormatManager.isDoubleNegative(amount)) {
                        cost += amount;
                    } else {
                        income += amount;
                    }
                }
            }

            result.put(currency, new double[]{cost, income});
        }
        return result;
    }

    private int getAccountPosById(int id) {
        for (int i = 0; i < accountList.size(); i++) {
            if (accountList.get(i).getId() == id) return i;
        }
        return -1;
    }

    private int getTransactionPosById(int id) {
        for (int i = 0; i < transactionList.size(); i++) {
            if (transactionList.get(i).getId() == id) return i;
        }
        return -1;
    }

    private int getDebtPosById(int id) {
        for (int i = 0; i < debtList.size(); i++) {
            if (debtList.get(i).getId() == id) return i;
        }
        return -1;
    }

    private int getTransactionCategoryIncomePosById(int id) {
        for (int i = 0; i < transactionCategoryIncomeList.size(); i++) {
            if (transactionCategoryIncomeList.get(i).getId() == id) return i;
        }
        return -1;
    }

    private int getTransactionCategoryExpensePosById(int id) {
        for (int i = 0; i < transactionCategoryExpenseList.size(); i++) {
            if (transactionCategoryExpenseList.get(i).getId() == id) return i;
        }
        return -1;
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/repository/memory/MemoryRepository.kt
package com.androidcollider.easyfin.common.repository.memory

import android.util.Log
import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager
import com.androidcollider.easyfin.common.models.*
import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
class MemoryRepository(
    private val numberFormatManager: NumberFormatManager,
    resourcesManager: ResourcesManager
) : Repository {
    private var accountList: MutableList<Account>? = null
    private var transactionList: MutableList<Transaction>? = null
    private var debtList: MutableList<Debt>? = null
    private var ratesArray: DoubleArray? = null
    private val currencyArray: Array<String>
    private var transactionCategoryIncomeList: MutableList<TransactionCategory>? = null
    private var transactionCategoryExpenseList: MutableList<TransactionCategory>? = null

    init {
        currencyArray = resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY)
    }

    override fun addNewAccount(account: Account): Single<Account> {
        return Single.fromCallable {
            accountList?.add(account)
            account
        }
    }

    override val allAccounts: Single<List<Account>>?
        get() {
            Log.d(TAG, "getAllAccounts")
            accountList?.let { return Single.just(it) }
            return null
        }

    override fun updateAccount(account: Account): Single<Account> {
        return Single.fromCallable {
            accountList?.let {
                val pos = it.indexOf(account)
                if (pos >= 0) it[pos] = account
            }
            account
        }
    }

    override fun deleteAccount(id: Int): Single<Boolean> {
        return Single.fromCallable {
            val pos = getAccountPosById(id)
            val b = pos != -1
            if (b) accountList?.removeAt(pos)
            b
        }
    }

    override fun transferBTWAccounts(
        idAccount1: Int,
        accountAmount1: Double,
        idAccount2: Int,
        accountAmount2: Double
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getAccountPosById(idAccount1)
            val b1 = pos != -1
            if (b1) {
                accountList!![pos].amount = accountAmount1
            }
            pos = getAccountPosById(idAccount2)
            val b2 = pos != -1
            if (b2) {
                accountList!![pos].amount = accountAmount2
            }
            b1 && b2
        }
    }

    override fun addNewTransaction(transaction: Transaction): Single<Transaction> {
        return Single.fromCallable {
            transactionList?.add(0, transaction)
            val pos = getAccountPosById(transaction.idAccount)
            if (pos != -1) {
                accountList!![pos].amount = transaction.accountAmount
            }
            transaction
        }
    }

    override val allTransactions: Single<List<Transaction>>?
        get() {
            Log.d(TAG, "getAllTransactions")
            transactionList?.let { return Single.just(it) }
            return null
        }

    override fun updateTransaction(transaction: Transaction): Single<Transaction> {
        return Single.fromCallable {
            var pos = getAccountPosById(transaction.idAccount)
            if (pos != -1) {
                accountList!![pos].amount = transaction.accountAmount
            }
            pos = transactionList!!.indexOf(transaction)
            if (pos >= 0) transactionList!![pos] = transaction
            transaction
        }
    }

    override fun updateTransactionDifferentAccounts(
        transaction: Transaction,
        oldAccountAmount: Double,
        oldAccountId: Int
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getAccountPosById(transaction.idAccount)
            val b1 = pos != -1
            if (b1) {
                accountList!![pos].amount = transaction.accountAmount
            }
            pos = getAccountPosById(oldAccountId)
            val b2 = pos != -1
            if (b2) {
                accountList!![pos].amount = oldAccountAmount
            }
            pos = transactionList!!.indexOf(transaction)
            val b3 = pos != -1
            if (b3) transactionList!![pos] = transaction
            b1 && b2 && b3
        }
    }

    override fun deleteTransaction(
        idAccount: Int,
        idTransaction: Int,
        amount: Double
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getTransactionPosById(idTransaction)
            var b = pos != -1
            if (b) {
                transactionList!!.removeAt(pos)
                pos = getAccountPosById(idAccount)
                b = pos != -1
                if (b) {
                    accountList!![pos].amount = accountList!![pos].amount - amount
                }
            }
            b
        }
    }

    override fun addNewDebt(debt: Debt): Single<Debt> {
        return Single.fromCallable {
            debtList!!.add(debt)
            val pos = getAccountPosById(debt.idAccount)
            if (pos != -1) {
                accountList!![pos].amount = debt.accountAmount
            }
            debt
        }
    }

    override val allDebts: Single<List<Debt>>?
        get() {
            Log.d(TAG, "getAllDebts")
            debtList?.let { return Single.just(it) }
            return null
        }

    override fun updateDebt(debt: Debt): Single<Debt> {
        return Single.fromCallable {
            var pos = getAccountPosById(debt.idAccount)
            if (pos != -1) {
                accountList!![pos].amount = debt.accountAmount
            }
            pos = debtList!!.indexOf(debt)
            if (pos >= 0) debtList!![pos] = debt
            debt
        }
    }

    override fun updateDebtDifferentAccounts(
        debt: Debt,
        oldAccountAmount: Double,
        oldAccountId: Int
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getAccountPosById(debt.idAccount)
            val b1 = pos != -1
            if (b1) {
                accountList!![pos].amount = debt.accountAmount
            }
            pos = getAccountPosById(oldAccountId)
            val b2 = pos != -1
            if (b2) {
                accountList!![pos].amount = oldAccountAmount
            }
            pos = debtList!!.indexOf(debt)
            val b3 = pos != -1
            if (b3) debtList!![pos] = debt
            b1 && b2 && b3
        }
    }

    override fun deleteDebt(
        idAccount: Int,
        idDebt: Int,
        amount: Double,
        type: Int
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getDebtPosById(idDebt)
            var b = pos != -1
            if (b) {
                debtList!!.removeAt(pos)
                pos = getAccountPosById(idAccount)
                b = pos != -1
                if (b) {
                    accountList!![pos].amount =
                        if (type == 1) accountList!![pos].amount - amount else accountList!![pos].amount + amount
                }
            }
            b
        }
    }

    override fun payFullDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getDebtPosById(idDebt)
            var b = pos != -1
            if (b) {
                debtList!!.removeAt(pos)
                pos = getAccountPosById(idAccount)
                b = pos != -1
                if (b) {
                    accountList!![pos].amount = accountAmount
                }
            }
            b
        }
    }

    override fun payPartOfDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int,
        debtAmount: Double
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getDebtPosById(idDebt)
            var b = pos != -1
            if (b) {
                debtList!![pos].amountCurrent = debtAmount
                pos = getAccountPosById(idAccount)
                b = pos != -1
                if (b) {
                    accountList!![pos].amount = accountAmount
                }
            }
            b
        }
    }

    override fun takeMoreDebt(
        idAccount: Int,
        accountAmount: Double,
        idDebt: Int,
        debtAmount: Double,
        debtAllAmount: Double
    ): Single<Boolean> {
        return Single.fromCallable {
            var pos = getDebtPosById(idDebt)
            var b = pos != -1
            if (b) {
                debtList!![pos].amountCurrent = debtAmount
                debtList!![pos].amountAll = debtAllAmount
                pos = getAccountPosById(idAccount)
                b = pos != -1
                if (b) {
                    accountList!![pos].amount = accountAmount
                }
            }
            b
        }
    }

    override fun getTransactionsStatistic(position: Int): Single<Map<String, DoubleArray>>? {
        return if (transactionList == null) null
        else Single.fromCallable { getTransactionsStatisticByPosition(position) }
    }

    override val accountsAmountSumGroupByTypeAndCurrency: Single<Map<String, DoubleArray>>?
        get() = if (accountList == null || debtList == null) null
        else Single.fromCallable { accountsSumGroupByTypeAndCurrency }

    override fun updateRates(ratesList: List<Rates>): Single<Boolean> {
        return Single.fromCallable {
            for (i in ratesArray!!.indices) {
                ratesArray!![i] = ratesList[i].ask
            }
            true
        }
    }

    override val rates: Single<DoubleArray>?
        get() {
            Log.d(TAG, "getRates")
            ratesArray?.let { return Single.just(it) }
            return null
        }

    override fun setAllAccounts(accountList: List<Account>): Single<Boolean> {
        return Single.fromCallable {
            this.accountList = ArrayList()
            this.accountList?.addAll(accountList)
            true
        }
    }

    override fun setAllTransactions(transactionList: List<Transaction>): Single<Boolean> {
        return Single.fromCallable {
            this.transactionList = ArrayList()
            this.transactionList?.addAll(transactionList)
            true
        }
    }

    override fun setAllDebts(debtList: List<Debt>): Single<Boolean> {
        return Single.fromCallable {
            this.debtList = ArrayList()
            this.debtList?.addAll(debtList)
            true
        }
    }

    override fun setRates(rates: DoubleArray): Single<Boolean> {
        return Single.fromCallable {
            ratesArray = DoubleArray(4)
            System.arraycopy(rates, 0, ratesArray!!, 0, rates.size)
            true
        }
    }

    override fun addNewTransactionIncomeCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory> {
        return Single.fromCallable {
            transactionCategoryIncomeList?.add(transactionCategory)
            transactionCategory
        }
    }

    override val allTransactionIncomeCategories: Single<List<TransactionCategory>>?
        get() = if (transactionCategoryIncomeList == null) null
        else Single.just(transactionCategoryIncomeList!!)

    override fun updateTransactionIncomeCategory(
        transactionCategory: TransactionCategory
    ): Single<TransactionCategory> {
        return Single.fromCallable {
            val pos = transactionCategoryIncomeList!!.indexOf(transactionCategory)
            if (pos >= 0) transactionCategoryIncomeList!![pos] = transactionCategory
            transactionCategory
        }
    }

    override fun deleteTransactionIncomeCategory(id: Int): Single<Boolean> {
        return Single.fromCallable {
            val pos = getTransactionCategoryIncomePosById(id)
            val b = pos != -1
            if (b) transactionCategoryIncomeList!!.removeAt(pos)
            b
        }
    }

    override fun setAllTransactionIncomeCategories(
        transactionCategoryList: List<TransactionCategory>
    ): Single<Boolean> {
        return Single.fromCallable {
            transactionCategoryIncomeList = ArrayList()
            transactionCategoryIncomeList!!.addAll(transactionCategoryList)
            true
        }
    }

    override fun addNewTransactionExpenseCategory(transactionCategory: TransactionCategory):
            Single<TransactionCategory> {
        return Single.fromCallable {
            transactionCategoryExpenseList?.add(transactionCategory)
            transactionCategory
        }
    }

    override val allTransactionExpenseCategories: Single<List<TransactionCategory>>?
        get() = if (transactionCategoryExpenseList == null) null
        else Single.just(transactionCategoryExpenseList!!)

    override fun updateTransactionExpenseCategory(
        transactionCategory: TransactionCategory
    ): Single<TransactionCategory> {
        return Single.fromCallable {
            val pos = transactionCategoryExpenseList!!.indexOf(transactionCategory)
            if (pos >= 0) transactionCategoryExpenseList!![pos] = transactionCategory
            transactionCategory
        }
    }

    override fun deleteTransactionExpenseCategory(id: Int): Single<Boolean> {
        return Single.fromCallable {
            val pos = getTransactionCategoryExpensePosById(id)
            val b = pos != -1
            if (b) transactionCategoryExpenseList!!.removeAt(pos)
            b
        }
    }

    override fun setAllTransactionExpenseCategories(
        transactionCategoryList: List<TransactionCategory>
    ): Single<Boolean> {
        return Single.fromCallable {
            transactionCategoryExpenseList = ArrayList()
            transactionCategoryExpenseList!!.addAll(transactionCategoryList)
            true
        }
    }

    private val accountsSumGroupByTypeAndCurrency: Map<String, DoubleArray>
        get() {
            Log.d(TAG, "getAccountsSumGroupByTypeAndCurrency")
            val results: MutableMap<String, DoubleArray> = HashMap()
            for (currency in currencyArray) {
                val accountListFilteredByCurrency = accountList!!
                    .filter { a: Account -> a.currency == currency }
                val result = DoubleArray(4)
                var accountSum: Double
                for (i in 0..2) {
                    accountSum = 0.0
                    val accountListFilteredByType = accountListFilteredByCurrency
                        .filter { a: Account -> a.type == i }
                    for (account in accountListFilteredByType) {
                        accountSum += account.amount
                    }
                    result[i] = accountSum
                }
                val debtListFilteredByCurrency = debtList!!
                    .filter { d: Debt -> d.currency == currency }
                var debtSum = 0.0
                var debtVal: Double
                for (debt in debtListFilteredByCurrency) {
                    debtVal = debt.amountCurrent
                    if (debt.type == 1) {
                        debtVal *= -1.0
                    }
                    debtSum += debtVal
                }
                result[3] = debtSum
                results[currency] = result
            }
            return results
        }

    private fun getTransactionsStatisticByPosition(position: Int): Map<String, DoubleArray> {
        Log.d(TAG, "getTransactionsStatisticByPosition")
        var period: Long = 0
        when (position) {
            1 -> period = DateConstants.DAY
            2 -> period = DateConstants.WEEK
            3 -> period = DateConstants.MONTH
            4 -> period = DateConstants.YEAR
            5 -> period = Long.MAX_VALUE
        }
        val result: MutableMap<String, DoubleArray> = HashMap()
        val currentTime = System.currentTimeMillis()
        var cost: Double
        var income: Double
        for (currency in currencyArray) {
            val transactionListFilteredByCurrency = transactionList!!
                .filter { t: Transaction -> t.currency == currency }
            cost = 0.0
            income = 0.0
            for (transaction in transactionListFilteredByCurrency) {
                val date = transaction.date
                val amount = transaction.amount
                if (currentTime > date && period >= currentTime - date) {
                    if (numberFormatManager.isDoubleNegative(amount)) {
                        cost += amount
                    } else {
                        income += amount
                    }
                }
            }
            result[currency] = doubleArrayOf(cost, income)
        }
        return result
    }

    private fun getAccountPosById(id: Int): Int {
        for (i in accountList!!.indices) {
            if (accountList!![i].id == id) return i
        }
        return -1
    }

    private fun getTransactionPosById(id: Int): Int {
        for (i in transactionList!!.indices) {
            if (transactionList!![i].id == id) return i
        }
        return -1
    }

    private fun getDebtPosById(id: Int): Int {
        for (i in debtList!!.indices) {
            if (debtList!![i].id == id) return i
        }
        return -1
    }

    private fun getTransactionCategoryIncomePosById(id: Int): Int {
        for (i in transactionCategoryIncomeList!!.indices) {
            if (transactionCategoryIncomeList!![i].id == id) return i
        }
        return -1
    }

    private fun getTransactionCategoryExpensePosById(id: Int): Int {
        for (i in transactionCategoryExpenseList!!.indices) {
            if (transactionCategoryExpenseList!![i].id == id) return i
        }
        return -1
    }

    companion object {
        private val TAG = MemoryRepository::class.java.simpleName
    }
}